### PR TITLE
Fix: Enable prefill phase key value caching of nemotron/minitron models

### LINKED ISSUE #34739: BUG : Modeling nemotron file does not cache key values even though
### System Info

huggingface-hub-0.26.2 
tokenizers-0.20.3 
transformers-4.47.0.dev0
Python 3.10.12
Driver Version: 535.129.03   
CUDA Version: 12.3 

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

## Description

You can run the below code to reproduce prefill key value caching problem of minitron models.
I used "nvidia/Minitron-8B-Base" model.

## Code 
```
from transformers import AutoModelForCausalLM, AutoTokenizer
import torch

# Load Minitron model and tokenizer from Hugging Face
model_name = "your-minitron-model-name"  # Replace with the actual Minitron model name
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(model_name)

# Set the model to evaluation mode
model.eval()

# Sample input text
input_text = "Hello, how are you?"

# Tokenize the input
input_ids = tokenizer(input_text, return_tensors="pt").input_ids

# First forward pass (prefill phase)
with torch.no_grad():
    outputs = model(input_ids, use_cache=True)  # Set use_cache=True
    logits = outputs.logits
    past_key_values = outputs.past_key_values

# Check the output
print("Logits shape:", logits.shape)
print("Number of layers in past_key_values:", len(past_key_values))
print("Shape of keys and values in the first layer:")
print("Key shape:", past_key_values[0][0].shape)
print("Value shape:", past_key_values[0][1].shape)

# Add new input to test cache utilization
new_input_text = " What about you?"
new_input_ids = tokenizer(new_input_text, return_tensors="pt").input_ids

# Pass the new input along with the previous key-value cache
with torch.no_grad():
    outputs_with_cache = model(new_input_ids, past_key_values=past_key_values, use_cache=True)

# Check results after caching
new_logits = outputs_with_cache.logits
new_past_key_values = outputs_with_cache.past_key_values

print("New logits shape:", new_logits.shape)
print("Number of layers in new past_key_values:", len(new_past_key_values))

```

### Expected behavior

**As-Is**
스크린샷 2024-11-15 오후 1 15 33

Past key value is 'Nonetype', which means the key value caches are not cached.

### PR DESCRIPTION
**Pull request by jeongin601**
# What does this PR do?

Fixes #34739

## Problem

Current implementation does not enable key value caching of nemotron and minitron models.
This problem can be checked by a quick example code that generates key and value caches.

## Modification
I modified the code to enabled key value caching while prefill phase, in reference to` 'modeling_llama.py'` file.

## Key value caching example code
```python
from transformers import AutoModelForCausalLM, AutoTokenizer
import torch

# Load Minitron model and tokenizer from Hugging Face
model_name = "your-minitron-model-name"  # Replace with the actual Minitron model name
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(model_name)

# Set the model to evaluation mode
model.eval()

# Sample input text
input_text = "Hello, how are you?"

# Tokenize the input
input_ids = tokenizer(input_text, return_tensors="pt").input_ids

# First forward pass (prefill phase)
with torch.no_grad():
    outputs = model(input_ids, use_cache=True)  # Set use_cache=True
    logits = outputs.logits
    past_key_values = outputs.past_key_values

# Check the output
print("Logits shape:", logits.shape)
print("Number of layers in past_key_values:", len(past_key_values))
print("Shape of keys and values in the first layer:")
print("Key shape:", past_key_values[0][0].shape)
print("Value shape:", past_key_values[0][1].shape)

# Add new input to test cache utilization
new_input_text = " What about you?"
new_input_ids = tokenizer(new_input_text, return_tensors="pt").input_ids

# Pass the new input along with the previous key-value cache
with torch.no_grad():
    outputs_with_cache = model(new_input_ids, past_key_values=past_key_values, use_cache=True)

# Check results after caching
new_logits = outputs_with_cache.logits
new_past_key_values = outputs_with_cache.past_key_values

print("New logits shape:", new_logits.shape)
print("Number of layers in new past_key_values:", len(new_past_key_values))
```
### As-Is Result 
스크린샷 2024-11-15 오후 1 15 33

Key value caching is not done.

### To-be Result
스크린샷 2024-11-15 오후 4 15 44

Key value caching is enabled

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker
Can you please check my modification? :)

### CODE REVIEW

**Comment by LysandreJik** [comment]
cc @ArthurZucker @gante @zucchini-nlp

File: src/transformers/models/nemotron/modeling_nemotron.py [hunk 1]
```diff
@@ -783,8 +783,22 @@ def forward(
...
+        if use_cache and not isinstance(past_key_values, Cache):
+            if past_key_values is None:
+                past_key_values = DynamicCache()
+            else:
+                past_key_values = DynamicCache.from_legacy_cache(past_key_values)
+                logger.warning_once(
+                    "We detected that you are passing `past_key_values` as a tuple of tuples. This is deprecated and "
+                    "will be removed in v4.47. Please convert your cache or use an appropriate `Cache` class "
+                    "(https://huggingface.co/docs/transformers/kv_cache#legacy-cache-format)"
+                )
```
Comment (zucchini-nlp) [inline src/transformers/models/nemotron/modeling_nemotron.py:795]: I don't think we need to add deprecation message for newly added models, we can support only new Cache objects

**Review by zucchini-nlp — COMMENTED** [review]
Thanks for adding this! Let's remove the deprecation warning, otherwise LGTM!

File: src/transformers/models/nemotron/modeling_nemotron.py (hunk 1 above)
Comment (jeongin601) [reply src/transformers/models/nemotron/modeling_nemotron.py]: Thanks for reviewing my code! I removed the deprecation warning. :)

Comment (zucchini-nlp) [reply src/transformers/models/nemotron/modeling_nemotron.py]: sorry if I wasn't clear, I mean totally removing support for the `tuple` format and this the `from_legacy_cache`

I'll ping the core maintainer after that for the final review :)

Comment (jeongin601) [reply src/transformers/models/nemotron/modeling_nemotron.py]: Oh, sorry I got it wrong. Now, I removed support for tuple shaped past_key_values. Is this what you meant?

File: src/transformers/models/nemotron/modeling_nemotron.py [hunk 2]
```diff
@@ -783,8 +783,15 @@ def forward(
         if inputs_embeds is None:
             inputs_embeds = self.embed_tokens(input_ids)
 
+        if use_cache and not isinstance(past_key_values, Cache):
+            if past_key_values is None:
+                past_key_values = DynamicCache()
```
Comment (ArthurZucker) [inline src/transformers/models/nemotron/modeling_nemotron.py:788]:
```suggestion
        if use_cache and past_key_values is None:
            past_key_values = DynamicCache()
```

**Review by ArthurZucker — APPROVED** [review]
One suggestion and we can merge!

**Comment by ArthurZucker** [comment]
BTW seem to be related to #34274

**Comment by jeongin601** [comment]
> One suggestion and we can merge!

I updated it! :) Thanks

**Comment by ydshieh** [comment]
Hi @jeongin601, thank you for this PR ❤️ !

It seems this PR has some regression: below is the 3 tests failing. Would you be up to take a look? You can find the job run log [here](https://github.com/huggingface/transformers/actions/runs/12022766734/job/33515773773).

In any case, thank you in advance.


```python
        "nemotron": {
            "single-gpu": [
                {
                    "test": "tests/models/nemotron/test_modeling_nemotron.py::NemotronModelTest::test_torchscript_output_attentions",
                    "commit": "318fe25f22a99ce1226f8d2aadc268b40f7e55af",
                    "pr_number": 34742,
                    "author": "jeongin601",
                    "merged_by": "zucchini-nlp"
                },
                {
                    "test": "tests/models/nemotron/test_modeling_nemotron.py::NemotronModelTest::test_torchscript_output_hidden_state",
                    "commit": "318fe25f22a99ce1226f8d2aadc268b40f7e55af",
                    "pr_number": 34742,
                    "author": "jeongin601",
                    "merged_by": "zucchini-nlp"
                },
                {
                    "test": "tests/models/nemotron/test_modeling_nemotron.py::NemotronModelTest::test_torchscript_simple",
                    "commit": "318fe25f22a99ce1226f8d2aadc268b40f7e55af",
                    "pr_number": 34742,
                    "author": "jeongin601",
                    "merged_by": "zucchini-nlp"
                }
            ]
        }
    },
```

**Comment by ydshieh** [comment]
@ArthurZucker @zucchini-nlp A kind remind: don't hesitate to ask for slow CI 🙂 - let's use the tools we have to make our life easier🙏

**Comment by ArthurZucker** [comment]
Ah it makes sense, torchscript does not support `DynamicCache` class !

**Comment by ArthurZucker** [comment]
(AFAIR)

**Comment by ydshieh** [comment]
ah ok. I will check than. But if we eventually move forward to `DynamicCache` and drop legacy cache, it would mean torchscript is not going to work for many models ..?

**Comment by ArthurZucker** [comment]
Yeah 👀 unless used with optimum!

**Pull request merged**